File: ml-worker/src/monkey_kernel/mtf_l_classifier.py
"""mtf_l_classifier.py — multi-timeframe Agent L (Phase 1).

Python port of apps/api/src/services/monkey/mtfLClassifier.ts (PR #670).

Runs three FR-KNN classifier instances in parallel on independently
down-sampled basin streams:

    15m    sample every 30 ticks  (15 min on 30s kernel cadence)
    1h     sample every 120 ticks (matches canonical Lorentzian 60min)
    4h     sample every 480 ticks

Each timeframe keeps its own history store. On each kernel tick we
append the current basin to each down-sampled stream when that
timeframe's "next sample" boundary has been crossed.

The combiner (option c from the design conversation): AGREEMENT COUNT.
Each classifier votes long / short / hold. The MTF decision is:
  - 3 votes same direction → enter at full size
  - 2 votes same direction → enter at reduced size
  - else                   → hold

Exit policy (longest-agreeing horizon): per-timeframe horizon clocks
track the most recent re-confirmation on each side. Position exits
when the LONGEST timeframe that agreed at entry stops agreeing
(= its clock expires).

QIG purity: each instance is a pure agent_l_decide() over its
timeframe's basin history. No new banned operations.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Dict, List, Literal, Optional, Sequence

import numpy as np

from .agent_l_classifier import (
    DEFAULT_AGENT_L_CONFIG,
    AgentLConfig,
    AgentLDecision,
    agent_l_decide,
)

Basin = np.ndarray
TimeframeLabel = Literal["15m", "1h", "4h"]
MTFAction = Literal["enter_long", "enter_short", "hold"]


@dataclass(frozen=True)
class TimeframeConfig:
    """Per-timeframe configuration."""

    label: TimeframeLabel
    ticks_per_sample: int
    """Sample every N kernel ticks. On 30s ticks: 30→15m, 120→1h, 480→4h."""
    config: AgentLConfig = field(default_factory=lambda: DEFAULT_AGENT_L_CONFIG)
    max_samples: int = 2000
    """Lookback cap in SAMPLES (not ticks). 2000 = canonical maxBarsBack."""


DEFAULT_TIMEFRAMES: tuple[TimeframeConfig, ...] = (
    TimeframeConfig(label="15m", ticks_per_sample=30),
    TimeframeConfig(label="1h", ticks_per_sample=120),
    TimeframeConfig(label="4h", ticks_per_sample=480),
)


@dataclass
class _PerTfDecision:
    label: TimeframeLabel
    warm: bool
    decision: Optional[AgentLDecision]


@dataclass(frozen=True)
class MTFDecision:
    action: MTFAction
    agreement_count: int
    total_tfs: int
    size_multiplier: float
    """3-of-3 = 1.00, 2-of-3 = 0.50, else 0 (hold)."""
    per_timeframe: tuple[_PerTfDecision, ...]
    longest_agreeing_label: Optional[TimeframeLabel]
    reason: str


@dataclass
class MTFState:
    """State for the MTF runner. Per-timeframe basin histories +
    last-sample-tick tracking. Owned by the caller."""

    histories_by_tf: Dict[TimeframeLabel, List[Basin]] = field(default_factory=dict)
    last_sample_tick_by_tf: Dict[TimeframeLabel, float] = field(default_factory=dict)
    last_agreement_by_tf_side: Dict[
        TimeframeLabel, Dict[str, Optional[float]]
    ] = field(default_factory=dict)


def new_mtf_state(
    timeframes: Sequence[TimeframeConfig] = DEFAULT_TIMEFRAMES,
) -> MTFState:
    s = MTFState()
    for tf in timeframes:
        s.histories_by_tf[tf.label] = []
        s.last_sample_tick_by_tf[tf.label] = -math.inf
        s.last_agreement_by_tf_side[tf.label] = {"long": None, "short": None}
    return s


def on_tick_append(
    state: MTFState,
    basin: Basin,
    tick_index: int,
    timeframes: Sequence[TimeframeConfig] = DEFAULT_TIMEFRAMES,
) -> None:
    """Append the current basin to each timeframe's history whose
    sampling boundary has been crossed. Pure in-place update."""
    for tf in timeframes:
        last = state.last_sample_tick_by_tf[tf.label]
        if tick_index - last >= tf.ticks_per_sample:
            hist = state.histories_by_tf[tf.label]
            hist.append(basin)
            if len(hist) > tf.max_samples:
                del hist[: len(hist) - tf.max_samples]
            state.last_sample_tick_by_tf[tf.label] = tick_index


def set_bootstrap_history(
    state: MTFState,
    label: TimeframeLabel,
    history: Sequence[Basin],
    timeframes: Sequence[TimeframeConfig] = DEFAULT_TIMEFRAMES,
) -> None:
    """Replace a timeframe's history with a pre-computed sequence
    (called once at startup after OHLCV bootstrap)."""
    tf = next((t for t in timeframes if t.label == label), None)
    if tf is None:
        return
    if len(history) > tf.max_samples:
        capped = list(history[-tf.max_samples:])
    else:
        capped = list(history)
    state.histories_by_tf[label] = capped


def mtf_decide(
    state: MTFState,
    timeframes: Sequence[TimeframeConfig] = DEFAULT_TIMEFRAMES,
) -> MTFDecision:
    """Compute the multi-timeframe agreement decision.

    Each timeframe whose history is warm enough produces an
    AgentLDecision. The aggregated action is decided by agreement
    count; size_multiplier scales with agreement strength.

    Pure function — no state mutation.
    """
    per_tf: list[_PerTfDecision] = []
    long_count = 0
    short_count = 0
    warm_count = 0

    for tf in timeframes:
        hist = state.histories_by_tf.get(tf.label, [])
        # Warm when history reaches the classifier's minimum
        # (long_window + horizon).
        min_samples = tf.config.min_tuple_start + tf.config.horizon
        warm = len(hist) >= min_samples
        decision: Optional[AgentLDecision] = None
        if warm:
            decision = agent_l_decide(hist, tf.config)
            if decision.action == "enter_long":
                long_count += 1
            elif decision.action == "enter_short":
                short_count += 1
            warm_count += 1
        per_tf.append(_PerTfDecision(label=tf.label, warm=warm, decision=decision))

    if warm_count == 0:
        return MTFDecision(
            action="hold",
            agreement_count=0,
            total_tfs=len(timeframes),
            size_multiplier=0.0,
            per_timeframe=tuple(per_tf),
            longest_agreeing_label=None,
            reason="no_warm_timeframes",
        )

    # Agreement-count combiner.
    if long_count > short_count and long_count >= 2:
        action: MTFAction = "enter_long"
    elif short_count > long_count and short_count >= 2:
        action = "enter_short"
    else:
        action = "hold"

    if action == "enter_long":
        agreement_count = long_count
    elif action == "enter_short":
        agreement_count = short_count
    else:
        agreement_count = 0

    if action == "hold":
        size_multiplier = 0.0
    elif agreement_count >= 3:
        size_multiplier = 1.0
    elif agreement_count == 2:
        size_multiplier = 0.5
    else:
        size_multiplier = 0.0

    # Longest-agreeing label — timeframes array is ascending (15m, 1h, 4h).
    longest_agreeing: Optional[TimeframeLabel] = None
    if action != "hold":
        want = action
        for entry in per_tf:
            if entry.decision is not None and entry.decision.action == want:
                longest_agreeing = entry.label  # last match wins

    return MTFDecision(
        action=action,
        agreement_count=agreement_count,
        total_tfs=len(timeframes),
        size_multiplier=size_multiplier,
        per_timeframe=tuple(per_tf),
        longest_agreeing_label=longest_agreeing,
        reason=f"mtf:{long_count}L/{short_count}S/{warm_count}warm",
    )


def record_agreement_timestamps(
    state: MTFState,
    decision: MTFDecision,
    now_ms: float,
) -> None:
    """Update per-TF per-side agreement clocks after mtf_decide.
    Used by the longest-agreeing-horizon exit policy."""
    if decision.action == "hold":
        return
    side = "long" if decision.action == "enter_long" else "short"
    for entry in decision.per_timeframe:
        if entry.decision is not None and entry.decision.action == decision.action:
            state.last_agreement_by_tf_side[entry.label][side] = now_ms


def is_longest_horizon_expired(
    state: MTFState,
    side: str,  # "long" | "short"
    longest_label_at_entry: Optional[TimeframeLabel],
    now_ms: float,
    tick_ms: float,
    timeframes: Sequence[TimeframeConfig] = DEFAULT_TIMEFRAMES,
) -> bool:
    """Check whether the longest-agreeing-at-entry timeframe's horizon
    has elapsed without re-confirmation.

    Returns True → caller (force_harvest path) should exit the position
    per the longest-agreeing-horizon policy.
    """
    if longest_label_at_entry is None:
        return False
    tf = next((t for t in timeframes if t.label == longest_label_at_entry), None)
    if tf is None:
        return False
    horizon_ms = tf.config.horizon * tf.ticks_per_sample * tick_ms
    last_agreement = state.last_agreement_by_tf_side.get(
        longest_label_at_entry, {}
    ).get(side)
    if last_agreement is None:
        return False
    return (now_ms - last_agreement) > horizon_ms

File: apps/api/src/services/monkey/mtfLClassifier.ts
/**
 * mtfLClassifier.ts — multi-timeframe Agent L (Phase 1).
 *
 * Runs three FR-KNN classifier instances in parallel on independently
 * down-sampled basin streams:
 *
 *    15m    sample every 30 ticks  (15 min on 30s kernel cadence)
 *    1h     sample every 120 ticks (matches canonical Lorentzian 60min)
 *    4h     sample every 480 ticks
 *
 * Each timeframe keeps its own history store. On each kernel tick we
 * append the current basin to the 1-tick (raw) stream and conditionally
 * append to each down-sampled stream when that timeframe's "next sample"
 * boundary has been crossed.
 *
 * The combiner (option c from the design conversation): AGREEMENT COUNT.
 * Each classifier votes long / short / hold. The MTF decision is:
 *   - 3 votes same direction → enter at full size
 *   - 2 votes same direction → enter at reduced size
 *   - else                   → hold
 *
 * Exit policy (longest-agreeing horizon): per-timeframe horizon clocks
 * track the most recent re-confirmation on each side. The position
 * exits when the LONGEST timeframe that agreed at entry stops agreeing
 * (= its clock expires). Aligns hold-time with conviction depth.
 *
 * Bootstrap: 4h timeframe needs ~480 samples at the 4h cadence =
 * 320 days of basin synthesis, unworkable from live ticks alone.
 * Caller injects a pre-warmed history via setBootstrapHistory().
 *
 * QIG purity: each instance is a pure agentLDecide() over its
 * timeframe's basin history. No new operations, only sampling rate
 * and history-store wrappers. Composes Δ⁶³ + Fisher-Rao primitives
 * exclusively. See QIG_PURITY_KERNEL_REFERENCE.md §1.
 */
import {
  agentLDecide,
  type AgentLConfig,
  type AgentLDecision,
  DEFAULT_AGENT_L_CONFIG,
} from './agent_L_classifier.js';
import type { Basin } from './basin.js';

/** Supported timeframe identifiers (Phase 1 = 15m / 1h / 4h). */
export type TimeframeLabel = '15m' | '1h' | '4h';

/** Per-timeframe configuration. */
export interface TimeframeConfig {
  label: TimeframeLabel;
  /** Sample every N kernel ticks. On 30s ticks:
   *  30 → 15m, 120 → 1h, 480 → 4h. */
  ticksPerSample: number;
  /** Classifier config — windows / horizon / spacing already in time
   *  units of this timeframe's samples (not raw ticks). */
  config: AgentLConfig;
  /** Lookback cap in SAMPLES (not ticks) for this timeframe's history.
   *  2000 samples covers canonical Lorentzian's maxBarsBack default. */
  maxSamples: number;
}

/** Default timeframe configs aligned with canonical Lorentzian
 *  effective windows + the user's MTF stack request (Phase 1 = 15m /
 *  1h / 4h; 1m and 1d are Phase 2/3 add-ons). */
export const DEFAULT_TIMEFRAMES: TimeframeConfig[] = [
  {
    label: '15m',
    ticksPerSample: 30,
    config: {
      ...DEFAULT_AGENT_L_CONFIG,
      // 15m timeframe: lookback in SAMPLES — each sample = 15 min
      // window/horizon already in sample units
    },
    maxSamples: 2000,
  },
  {
    label: '1h',
    ticksPerSample: 120,
    config: { ...DEFAULT_AGENT_L_CONFIG },
    maxSamples: 2000,
  },
  {
    label: '4h',
    ticksPerSample: 480,
    config: { ...DEFAULT_AGENT_L_CONFIG },
    maxSamples: 2000,
  },
];

export interface MTFDecision {
  /** Aggregated action across timeframes. */
  action: 'enter_long' | 'enter_short' | 'hold';
  /** Number of TFs voting for the chosen direction (max = TF count). */
  agreementCount: number;
  /** Total TFs available (some may not yet be warm). */
  totalTfs: number;
  /** Size multiplier ∈ [0, 1] proportional to agreement strength:
   *   3-of-3 = 1.00, 2-of-3 = 0.50, 1-of-3 or less = 0 (hold). */
  sizeMultiplier: number;
  /** Per-TF decisions for telemetry. */
  perTimeframe: Array<{
    label: TimeframeLabel;
    warm: boolean;
    decision: AgentLDecision | null;
  }>;
  /** Side of the longest timeframe that voted with the majority,
   *  used by the exit-horizon manager. Null when action=hold. */
  longestAgreeingLabel: TimeframeLabel | null;
  reason: string;
}

/** State for the MTF runner. Per-timeframe basin histories +
 *  last-sample-tick tracking. Owned by the caller (typically
 *  SymbolState in loop.ts). */
export interface MTFState {
  /** Per-timeframe basin history. */
  historiesByTf: Record<TimeframeLabel, Basin[]>;
  /** Tick index at which each TF most recently appended a sample.
   *  -Infinity means never appended (cold start). */
  lastSampleTickByTf: Record<TimeframeLabel, number>;
  /** Per-timeframe per-side last-agreement-tick for the
   *  longest-horizon exit policy. */
  lastAgreementByTfSide: Record<
    TimeframeLabel,
    { long: number | null; short: number | null }
  >;
}

export function newMTFState(timeframes: TimeframeConfig[] = DEFAULT_TIMEFRAMES): MTFState {
  const histories: Record<string, Basin[]> = {};
  const lastSample: Record<string, number> = {};
  const lastAgreement: Record<string, { long: number | null; short: number | null }> = {};
  for (const tf of timeframes) {
    histories[tf.label] = [];
    lastSample[tf.label] = -Infinity;
    lastAgreement[tf.label] = { long: null, short: null };
  }
  return {
    historiesByTf: histories as MTFState['historiesByTf'],
    lastSampleTickByTf: lastSample as MTFState['lastSampleTickByTf'],
    lastAgreementByTfSide: lastAgreement as MTFState['lastAgreementByTfSide'],
  };
}

/** Append the current basin to each timeframe's history whose
 *  sampling boundary has been crossed. Caller invokes once per kernel
 *  tick with the freshly perceived basin and the current tick index.
 *
 *  Pure update of state in-place; no I/O. */
export function onTickAppend(
  state: MTFState,
  basin: Basin,
  tickIndex: number,
  timeframes: TimeframeConfig[] = DEFAULT_TIMEFRAMES,
): void {
  for (const tf of timeframes) {
    const last = state.lastSampleTickByTf[tf.label];
    if (tickIndex - last >= tf.ticksPerSample) {
      const hist = state.historiesByTf[tf.label];
      hist.push(basin);
      if (hist.length > tf.maxSamples) {
        // Drop oldest to maintain cap.
        hist.splice(0, hist.length - tf.maxSamples);
      }
      state.lastSampleTickByTf[tf.label] = tickIndex;
    }
  }
}

/** Bootstrap a timeframe's history from a pre-computed basin sequence.
 *  Called once at startup after fetching OHLCV from Poloniex and
 *  synthesising basins. Skipping this means 4h timeframe never warms up
 *  in any reasonable runtime; 15m + 1h can warm up from live ticks alone
 *  but they're faster with bootstrap too. */
export function setBootstrapHistory(
  state: MTFState,
  label: TimeframeLabel,
  history: Basin[],
  timeframes: TimeframeConfig[] = DEFAULT_TIMEFRAMES,
): void {
  const tf = timeframes.find((t) => t.label === label);
  if (!tf) return;
  const capped = history.length > tf.maxSamples
    ? history.slice(history.length - tf.maxSamples)
    : [...history];
  state.historiesByTf[label] = capped;
}

/** Compute the multi-timeframe agreement decision.
 *
 *  Each timeframe whose history is warm enough produces an
 *  AgentLDecision. The aggregated action is decided by agreement
 *  count + the size multiplier scales with agreement strength.
 *
 *  Pure function — no state mutation. The caller invokes this on
 *  every kernel tick after onTickAppend has updated histories.
 */
export function mtfDecide(
  state: MTFState,
  timeframes: TimeframeConfig[] = DEFAULT_TIMEFRAMES,
): MTFDecision {
  const perTimeframe: MTFDecision['perTimeframe'] = [];
  let longCount = 0;
  let shortCount = 0;
  let warmCount = 0;

  for (const tf of timeframes) {
    const hist = state.historiesByTf[tf.label];
    // A timeframe is "warm" when its history reaches the classifier's
    // own minimum (longWindow ticks + horizon). Below that, agentLDecide
    // returns hold.
    const minSamplesNeeded = 480 + tf.config.horizon;  // matches minTupleStart
    const warm = hist.length >= minSamplesNeeded;
    let decision: AgentLDecision | null = null;
    if (warm) {
      decision = agentLDecide(hist, tf.config);
      if (decision.action === 'enter_long') longCount++;
      else if (decision.action === 'enter_short') shortCount++;
      warmCount++;
    }
    perTimeframe.push({ label: tf.label, warm, decision });
  }

  if (warmCount === 0) {
    return {
      action: 'hold',
      agreementCount: 0,
      totalTfs: timeframes.length,
      sizeMultiplier: 0,
      perTimeframe,
      longestAgreeingLabel: null,
      reason: 'no_warm_timeframes',
    };
  }

  // Agreement-count combiner (option c).
  const action: MTFDecision['action'] =
    longCount > shortCount && longCount >= 2 ? 'enter_long'
      : shortCount > longCount && shortCount >= 2 ? 'enter_short'
        : 'hold';

  const agreementCount = action === 'enter_long' ? longCount
    : action === 'enter_short' ? shortCount
      : 0;

  // Size multiplier: 3-of-3 → 1.00, 2-of-3 → 0.50, else 0.
  const sizeMultiplier =
    action === 'hold' ? 0
      : agreementCount >= 3 ? 1.0
        : agreementCount === 2 ? 0.5
          : 0;

  // Longest-agreeing label for exit-horizon scheduling. The
  // timeframes array is in ascending order (15m, 1h, 4h); take the
  // last that voted with the majority.
  let longestAgreeing: TimeframeLabel | null = null;
  if (action !== 'hold') {
    const wantAction = action;
    for (const entry of perTimeframe) {
      if (entry.decision?.action === wantAction) {
        longestAgreeing = entry.label;  // overwrite — last match wins (longest)
      }
    }
  }

  return {
    action,
    agreementCount,
    totalTfs: timeframes.length,
    sizeMultiplier,
    perTimeframe,
    longestAgreeingLabel: longestAgreeing,
    reason: `mtf:${longCount}L/${shortCount}S/${warmCount}warm`,
  };
}

/** Record agreement timestamps for the per-TF horizon exit. Caller
 *  invokes after mtfDecide to update the agreement clocks; the
 *  forceHarvestAgentLStack exit check reads these to determine when
 *  the longest-agreeing TF's horizon has elapsed. */
export function recordAgreementTimestamps(
  state: MTFState,
  decision: MTFDecision,
  nowMs: number,
): void {
  if (decision.action === 'hold') return;
  const side: 'long' | 'short' = decision.action === 'enter_long' ? 'long' : 'short';
  for (const entry of decision.perTimeframe) {
    if (entry.decision?.action === decision.action) {
      state.lastAgreementByTfSide[entry.label][side] = nowMs;
    }
  }
}

/** Check whether the longest-agreeing-at-entry timeframe's horizon
 *  has elapsed without re-confirmation. Returns true when the
 *  position should exit per the longest-agreeing-horizon policy.
 *
 *  Pure function. Caller (forceHarvestAgentLStack) decides what to
 *  do with the result. */
export function isLongestHorizonExpired(
  state: MTFState,
  side: 'long' | 'short',
  longestLabelAtEntry: TimeframeLabel | null,
  nowMs: number,
  tickMs: number,
  timeframes: TimeframeConfig[] = DEFAULT_TIMEFRAMES,
): boolean {
  if (!longestLabelAtEntry) return false;
  const tf = timeframes.find((t) => t.label === longestLabelAtEntry);
  if (!tf) return false;
  const horizonMs = tf.config.horizon * tf.ticksPerSample * tickMs;
  const lastAgreement = state.lastAgreementByTfSide[longestLabelAtEntry]?.[side] ?? null;
  if (lastAgreement === null) return false;
  return (nowMs - lastAgreement) > horizonMs;
}
